drivers/at24cxxx/at24cxxx.c

```diff
@@ -0,0 +1,257 @@
+/*
+ * Copyright (C) 2019 Otto-von-Guericke-Universität Magdeburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_at24cxxx
+ * @{
+ *
+ * @file
+ * @brief       Device driver implementation for the AT24C128 / AT24C256 EEPROM units.
+ *
+ * @author      Fabian Hüßler <fabian.huessler@ovgu.de>
+ * @}
+ */
+
+#include <stdio.h>
+#include <errno.h>
+#include <string.h>
+
+#define ENABLE_DEBUG  (0)
+#include "debug.h"
+#include "assert.h"
+#include "xtimer.h"
+#include "byteorder.h"
+
+#include "include/at24cxxx_params.h"
+#include "include/at24cxxx_defines.h"
+#include "at24cxxx.h"
+
+#ifndef min
+#define min(a, b) ((a) > (b) ? (b) : (a))
+#endif
+
+static int32_t _read(const at24cxxx_t *dev, uint16_t pos, void *data,
+                     size_t len)
+{
+    int check;
+    uint8_t polls = AT24CXXX_MAX_POLLS;
+
+    while ((check =
+                i2c_read_regs(dev->params.i2c, dev->params.dev_addr, pos, data,
+                              len,
+                              I2C_REG16)) == -ENXIO) {
+        if (--polls == 0) {
+            break;
+        }
+        xtimer_usleep(AT24CXXX_POLL_DELAY_US);
+    }
+    DEBUG("_read::i2c_read_regs check: %d; polls: %d\n", check, polls);
+    return check < 0 ? check : (int32_t)len;
+}
+
+static int32_t _write(const at24cxxx_t *dev, uint16_t pos, const void *data,
+                      size_t len)
+{
+    int check = 0;
+    size_t wr = 0;
+    size_t chunk_len;
+
+    while (!check && (chunk_len = len - wr)) {
+        const uint8_t *chunk_data = ((const uint8_t *)data) + wr;
+        chunk_len =
+            min(chunk_len,
+                AT24CXXX_PAGE_SIZE - (pos & (AT24CXXX_PAGE_SIZE - 1)));
+        uint8_t polls = AT24CXXX_MAX_POLLS;
+        while ((check =
+                    i2c_write_regs(dev->params.i2c, dev->params.dev_addr, pos,
+                                   chunk_data,
+                                   chunk_len,
+                                   I2C_REG16)) == -ENXIO) {
+            if (--polls == 0) {
+                break;
+            }
+            xtimer_usleep(AT24CXXX_POLL_DELAY_US);
+        }
+        DEBUG("_write::i2c_write_regs check: %d; polls: %d\n", check, polls);
+        if (!check) {
+            wr += chunk_len;
+            pos += chunk_len;
+        }
+    }
+    return wr > 0 ? (int32_t)wr : check;
+}
+
+static int32_t _set(const at24cxxx_t *dev, uint16_t pos, uint8_t val,
+                    size_t len)
+{
+    int check = 0;
+    size_t wr = 0;
+    size_t chunk_len;
+    uint8_t set_buffer[AT24CXXX_PAGE_SIZE];
```

Comment:
I link this approach, as this will not only fix the issue with the STM32F0/F3/F7/L0/L4 boards, but also make more efficient use DMA. However, on some boards 64B of stack could easily overflow the stack.

How about adding something like

``` C
#ifndef AT24CXXX_SET_BUF_SIZE
#define AT24CXXX_SET_BUF_SIZE AT24CXXX_PAGE_SIZE
#endif
```
? Then people having issues could lower the buffer size to work for their use case.

Reply:
I once had the buffersize configurable for exactly that reason. However, I am gonna change it back then.